### PR TITLE
Bump Spot SDK from `4.0.2` to `4.1.0`

### PR DESCRIPTION
**Pull request by mhidalgo-bdai**
## Change Overview

Precisely what the title says. This patch updates both C++ and Python SDKs, as well as `bosdyn_msgs`  for ROS 2 integration.

## Testing Done

Relying on CI and downstream tests.

### CODE REVIEW

**Comment by mhidalgo-bdai** [comment]
Argh, the `bosdyn_msgs` bundle install assumes `wget` is available :facepalm:

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -30,27 +30,27 @@ RUN DEBIAN_FRONTEND=noninteractive apt-get update -q && \
 RUN source "/opt/ros/${ROS_DISTRO}/setup.bash"
 
 # Install bosdyn_msgs package
```
Comment (amessing-bdai) [inline .devcontainer/Dockerfile:32]:
```suggestion
ARG SPOT_SDK_VERSION="4.1.0"

# Install bosdyn_msgs package
```

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -30,27 +30,27 @@ RUN DEBIAN_FRONTEND=noninteractive apt-get update -q && \
...
-RUN curl -sL https://github.com/bdaiinstitute/bosdyn_msgs/releases/download/4.0.2/ros-humble-bosdyn_msgs_4.0.2-jammy_amd64.run --output /tmp/ros-humble-bosdyn_msgs_4.0.2-jammy_amd64.run --silent \
-  && chmod +x /tmp/ros-humble-bosdyn_msgs_4.0.2-jammy_amd64.run \
-  && ((yes || true) | /tmp/ros-humble-bosdyn_msgs_4.0.2-jammy_amd64.run) \
-  && rm /tmp/ros-humble-bosdyn_msgs_4.0.2-jammy_amd64.run
+RUN curl -sL https://github.com/bdaiinstitute/bosdyn_msgs/releases/download/4.1.0/ros-humble-bosdyn_msgs_4.1.0-jammy_amd64.run --output /tmp/ros-humble-bosdyn_msgs_4.1.0-jammy_amd64.run --silent \
```
Comment (amessing-bdai) [inline .devcontainer/Dockerfile:33]:
```suggestion
RUN curl -sL https://github.com/bdaiinstitute/bosdyn_msgs/releases/download/${SPOT_SDK_VERSION}/ros-humble-bosdyn_msgs_${SPOT_SDK_VERISON}-jammy_amd64.run --output /tmp/ros-humble-bosdyn_msgs_${SPOT_SDK_VERSION}-jammy_amd64.run --silent \
```

**Comment by khughes-bdai** [comment]
I just put up this PR in spot_wrapper https://github.com/bdaiinstitute/spot_wrapper/pull/143 so that these two repos can stay in sync -- the install_spot_ros2.sh script here points to spot_wrapper's requirements.txt, which could cause issues for external users

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (mhidalgo-bdai) [reply .devcontainer/Dockerfile]: Done in https://github.com/bdaiinstitute/spot_ros2/pull/495/commits/88f0b6d9cd1abeffe4d8d6f5d87c1f926ed25684.

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (mhidalgo-bdai) [reply .devcontainer/Dockerfile:35]: Done in https://github.com/bdaiinstitute/spot_ros2/pull/495/commits/88f0b6d9cd1abeffe4d8d6f5d87c1f926ed25684.

**Comment by khughes-bdai** [comment]
minor request, but can we update the README as a part of this PR to reflect this new sdk version?

**Review by tcappellari-bdai — APPROVED** [review]

**Comment by mhidalgo-bdai** [comment]
This is getting harder than expected. Spot C++ SDK 4.1.0 libraries seem to depend (yet don't declare the dependency) on a Protobuf version that is newer than the one that is available in Ubuntu 22.04. Static linkage might still work if there were no API breaks between Protobuf versions, but dynamic linkage won't. CI fails (and I can reproduce) to dynamically load `libbosdyn_api.so` on missing `google::protobuf::Message::CopyWithSourceCheck` symbols.

We should report this upstream.

**Comment by mhidalgo-bdai** [comment]
@tcappellari-bdai @khughes-bdai FYI https://github.com/bdaiinstitute/spot_ros2/pull/495/commits/6ad3c15e159bdd9792aece923eec6fc176b2356c. The OS distributed version of Protobuf (3.12.4) does not have `protobuf::RepeatedField<T>::Assign()`.

File: spot_driver/CMakeLists.txt [hunk 3]
```diff
@@ -85,7 +86,7 @@ target_include_directories(spot_api PUBLIC
   $<INSTALL_INTERFACE:include>
 )
 
-target_link_libraries(spot_api PUBLIC bosdyn::bosdyn_client_static)
+target_link_libraries(spot_api PUBLIC bosdyn::bosdyn_client)
```
Comment (khughes-bdai) [inline spot_driver/CMakeLists.txt:89]: Is this same change also needed for the ros2 control package?

File: spot_driver/CMakeLists.txt [hunk 4]
```diff
@@ -1,7 +1,5 @@
 cmake_minimum_required(VERSION 3.22)
 
-list(APPEND CMAKE_PREFIX_PATH /opt/spot-cpp-sdk)
-
```
Comment (khughes-bdai) [inline spot_driver/CMakeLists.txt:4]: this line is also in the spot_ros2_control cmake and probably could also be removed if it's not needed anymore

File: spot_driver/CMakeLists.txt (hunk 3 above)
Comment (mhidalgo-bdai) [reply spot_driver/CMakeLists.txt:89]: Good catch. Done in https://github.com/bdaiinstitute/spot_ros2/pull/495/commits/93d3b304feac6f10fa26013c81111c4f6483acb5.

File: spot_driver/CMakeLists.txt (hunk 4 above)
Comment (mhidalgo-bdai) [reply spot_driver/CMakeLists.txt:4]: Done in https://github.com/bdaiinstitute/spot_ros2/pull/495/commits/93d3b304feac6f10fa26013c81111c4f6483acb5.

**Review by khughes-bdai — APPROVED** [review]
LGTM! manually ran install script to successfully upgrade packages on my laptop, and successfully ran tests on robot in both `spot_driver` and `spot_ros2_control`

**Comment by mhidalgo-bdai** [comment]
### Merge activity

* **Oct 17, 5:22 PM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/bdaiinstitute/spot_ros2/495).

**Pull request merged**